File: Creational/Builder/cpp/builder.cpp
#include "builder.hpp"
#include <iostream>

using namespace builder_pattern;

/**
* @file builder.cpp
* @brief Implementation of the Builder Pattern.
*
* This file contains the implementation details of the `Product`, `ConcreteBuilder`,
* and `Director` classes, following the Builder Pattern to construct complex objects.
*/

// Product implementation

void Product::setPartA(const std::string& part) {
	this->partA = part;
}

void Product::setPartB(const std::string& part) {
	this->partB = part;
}

/**
* @brief Provides a string representation of the product.
*
* Combines the details of Part A and Part B into a formatted string
* that describes the constructed product.
*
* @return A string representation of the product's parts.
*/
std::string Product::toString() const {
	return "Product [PartA: " + partA + ", PartB: " + partB + "]";
}

/**
* @brief Overloads the output stream operator for the `Product` class.
*
* Allows the details of the `Product` to be printed to an output stream.
*
* @param os The output stream to write to.
* @param product The product whose details are printed.
* @return The output stream with appended product details.
*/
namespace builder_pattern { 
	std::ostream& operator<<(std::ostream& os, const Product& product) {
		os << product.toString();
		return os;
	}
}

// ConcreteBuilder implementation

/**
* @brief Constructs a new `ConcreteBuilder` and initializes a product.
*
* Initializes the builder with a fresh `Product` instance for construction.
*/
ConcreteBuilder::ConcreteBuilder() {
	product = std::make_shared<Product>();
}

/**
* @brief Builds Part A of the product.
*
* Sets Part A of the product to a specific value defined by the builder.
*/
void ConcreteBuilder::buildPartA() {
	product->setPartA("Part A built by ConcreteBuilder");
}

/**
* @brief Builds Part B of the product.
*
* Sets Part B of the product to a specific value defined by the builder.
*/
void ConcreteBuilder::buildPartB() {
	product->setPartB("Part B built by ConcreteBuilder");
}

/**
* @brief Retrieves the constructed product.
*
* Returns the fully constructed product after all parts have been built.
*
* @return A shared pointer to the constructed `Product`.
*/
std::shared_ptr<Product> ConcreteBuilder::getResult() {
	return product;
}

// Director implementation

Director::Director(Builder* builder) : builder(builder) {}

/**
* @brief Constructs the product by invoking the builder's methods in sequence.
*
* Orchestrates the construction process by calling the builder's methods
* in a predefined order.
*/
void Director::construct() {
	builder->buildPartA();
	builder->buildPartB();
}

/**
* @brief Retrieves the constructed product from the builder.
*
* After the construction process is complete, this method retrieves
* the final constructed product from the builder.
*
* @return A shared pointer to the constructed `Product`.
*/
std::shared_ptr<Product> Director::getProduct() {
	return builder->getResult();
}

File: Creational/Builder/python/builder.py
from abc import ABC, abstractmethod

class Product:
	"""
	Represents the complex product being built.

	The Product class contains attributes `part_a` and `part_b`, which represent
	the individual parts of the product. It provides methods to set these parts
	and a string representation to display the constructed product.
	"""

	def __init__(self):
		"""
		Initializes the Product instance.

		Sets the attributes `part_a` and `part_b` to None, which represent the
		individual parts of the product.
		"""
		self.part_a = None  # Represents Part A of the product
		self.part_b = None  # Represents Part B of the product

	def set_part_a(self, part_a):
		"""
		Sets Part A of the product.

		:param part_a: A string representing Part A of the product.
		"""
		self.part_a = part_a

	def set_part_b(self, part_b):
		"""
		Sets Part B of the product.

		:param part_b: A string representing Part B of the product.
		"""
		self.part_b = part_b

	def __str__(self):
		"""
		Returns a string representation of the product.

		:return: A string describing the product's parts.
		"""
		return f"Product [PartA: {self.part_a}, PartB: {self.part_b}]"


class Builder(ABC):
	"""
	Abstract interface for constructing a Product.

	Defines methods for building different parts of a product and retrieving
	the final constructed result.
	"""

	@abstractmethod
	def build_part_a(self):
		"""Builds Part A of the product."""
		pass

	@abstractmethod
	def build_part_b(self):
		"""Builds Part B of the product."""
		pass

	@abstractmethod
	def get_result(self) -> Product:
		"""
		Retrieves the constructed product.

		:return: The constructed Product instance.
		"""
		pass


class ConcreteBuilder(Builder):
	"""
	Concrete implementation of the Builder interface.

	Constructs a specific type of product by implementing the building steps
	defined in the Builder interface.
	"""

	def __init__(self):
		"""
		Initializes a new instance of the ConcreteBuilder class.

		Initializes the builder with a new instance of the Product class.
		"""
		self.product = Product()

	def build_part_a(self):
		"""
		Builds Part A of the product.

		Sets Part A to a specific value defined by this builder.
		"""
		self.product.set_part_a("Part A built by ConcreteBuilder")

	def build_part_b(self):
		"""
		Builds Part B of the product.

		Sets Part B to a specific value defined by this builder.
		"""
		self.product.set_part_b("Part B built by ConcreteBuilder")

	def get_result(self) -> Product:
		"""
		Retrieves the constructed product.

		:return: The fully constructed Product instance.
		"""
		return self.product


class Director:
	"""
	Orchestrates the construction process.

	The Director class uses a Builder to construct a Product in a predefined sequence.
	"""

	def __init__(self, builder: Builder):
		"""
		Initializes the Director with a builder.

		:param builder: The Builder instance to use for constructing the product.
		"""
		self.builder = builder

	def construct(self):
		"""
		Constructs the product by invoking the builder's methods in sequence.
		"""
		self.builder.build_part_a()
		self.builder.build_part_b()

	def get_product(self):
		"""
		Retrieves the constructed product from the builder.

		:return: The constructed Product instance.
		"""
		return self.builder.get_result()
